pkg/cdi/spec-dirs_test.go

```diff
@@ -242,8 +242,12 @@ func updateTestDir(t *testing.T, tmp string, dirs map[string]map[string]string)
 		}
 		for file, data := range content {
 			file := filepath.Join(dir, file)
-			if err := ioutil.WriteFile(file, []byte(data), 0644); err != nil {
-				return errors.Wrapf(err, "failed to write file %q", file)
+			tmp := file + ".tmp"
+			if err := ioutil.WriteFile(tmp, []byte(data), 0644); err != nil {
+				return errors.Wrapf(err, "failed to write file %q", tmp)
+			}
+			if err := os.Rename(tmp, file); err != nil {
```

Comment:
A subsequent commit at the tip of [this tree](https://github.com/klihub/container-device-interface/commits/devel/spec-write), which implements a simple API for writing Spec files, mirrors this same behavior.

The problem with first testing for the existence of a file and only then creating/writing it is that this is not atomic. There is a time window between the check and the creation operations where another process can create the exact file which was verified not to exist. If we want the API to provide means for preventing the replacement of an existing file we can do that by 
  - letting the caller explicitly indicate whether overwrite is desired (`Spec.Write(replace bool) error`)
  - using `[golang.org/x/sys/]unix.Renameat2(tmp, file, unix.RENAME_NOREPLACE)` for atomic check/replace

I can change the implementation if this is considered useful. @kad @elezar @bart0sh, WDYT ? 

So far this has not been a concern mostly due to the Spec file structure which syntactically limits a single Spec file to declare only instances of a single vendor/device-type combo. This implies a strong vendor-ownership of every Spec file, which then hopefully gives a strong hint for vendors/implementations to namespace their Spec files with a unique vendor-specific prefix.

Reply:
fair.. if the rename fails due to exist.. should we "fail" or resolve

Reply:
I think we should "fail"

Assuming that the caller can determine the desired filename, I would vote for keeping the API implementation simpler in that if a file already exists at the target we would error out (whether this is done atomically or not). The caller can then decide whether overwriting the file is the desired behaviour -- removing the existing file explicitly.

Reply:
With that said, since this is currently test code we can address these questions in the upcoming API additions for writing spec files. Would that be reasonable from your perspective @mikebrow?

Reply:
@klihub
+1 for letting the caller explicitly indicate whether overwrite is desired 

not sure if we need to use Renameat2 as some filesystems might not support it
Can we simply use os.OpenFile with O_CREATE when caller doesn't want to overwrite it?

Reply:
> I think we should "fail"
> 
> Assuming that the caller can determine the desired filename, I would vote for keeping the API implementation simpler in that if a file already exists at the target we would error out (whether this is done atomically or not). The caller can then decide whether overwriting the file is the desired behaviour -- removing the existing file explicitly.

@elezar Okay, so the primary consumer for the Spec Write() interface IMO is a CDI-enabled K8s device plugin, for instance an NVidia GPU device plugin. How would the logic for generating/updating/removing CDI Spec file entries/Spec files for that device plugin would look like wrt. K8s/DP events (device allocation and in the future potential device release notification) ? My assumption/target was that the device plugin would

  - (prefer to) use a single Spec file (clear ownership of spec files vs. device plugins)
  - atomically update the Spec file for every plugin state change (allocation, some kind of release notification) with the new content reflecting the updated state

This was the major reason I wanted to add the Spec Write interface and with the semantics I proposed.

But this might be a total misconception on my side. Therefore I think it would be a good idea if you guys (@elezar @bart0sh ) could think this through to understand whether there are any details in the typical/probable DP-generated Spec file lifecycle that a DP could easily get wrong. If there is, then it would be a good idea to provide a single correct implementation, with the right semantics, in the CDI package, so that DP authors could then use that.

Reply:
That is a fair point. I wasn't really considering the device plugin use case at this stage. I was thinking from a CLI perspective where I (as a device vendor) want to provide a utility to generate spec files for available devices that can be run from the command line in response to some package installation. In this case I want to be able to provide feedback to the user if something unexpected happens.

With that said, I think having a flag `replace` / `overwrite` that is passed to the `Write` method would be able to handle both cases. Meaning that it would fail if `replace` is not `true`.

May I suggest that we create the PR that you are discussing with the addition of this API and then move this discussion there. Although it is valuable, I think these nuanced discussions are out of scope for this particular test utility. Once we are working with the actual API we may also be able to reason about the changes better and provide usage examples.

Reply:
Yes, I agree that this definitely belongs to a separate PR. Just wanted to bring up that PoV while we were at it, since/once the topic came up. I can file a Spec Write API PR and we can sort out the details there.

Reply:
I filed that Spec Write API PR as #64. It is marked as draft because one of its test cases will need a small update once this auto-refresh PR is merged.